src/ray/raylet/node_manager.cc

```diff
@@ -2560,6 +2632,73 @@ void NodeManager::RecordMetrics() {
   object_directory_->RecordMetrics(duration_ms);
 }
 
+void NodeManager::Update(std::shared_ptr<const syncer::RaySyncMessage> message) {
+  if (message->component_id() == syncer::RayComponentId::RESOURCE_MANAGER) {
+    rpc::ResourcesData data;
+    data.ParseFromString(message->sync_message());
+    NodeID node_id = NodeID::FromBinary(data.node_id());
+    UpdateResourceUsage(node_id, data);
+  } else if (message->component_id() == syncer::RayComponentId::COMMANDS) {
+    rpc::ResourcesData data;
+    data.ParseFromString(message->sync_message());
+    if (data.should_global_gc()) {
+      should_local_gc_ = true;
+    }
+  }
+}
+
+std::optional<syncer::RaySyncMessage> NodeManager::Snapshot(
+    int64_t after_version, syncer::RayComponentId component_id) const {
+  // Right now snapshot is put in NodeManager which in long-term, them should
+  // be put into each component directly.
+  // The current blocker of doing this is some fields in NodeManager
+  // is being used.
+  // TODO(iycheng): Move the logic into the components directly.
```

Comment:
> should these two functions belong to cluster_task_manager_/cluster_resource_scheduler_ ?

@scv119  to answer your question, yes, they should. But there are some fields missing in the component, I put it here for now. For example, `resource_deadlock_warned_` is not there, self_node_id_ is not there and ...

I think a minor refactor is needed for this one which I'll do in 6.2 if you are ok with that.